8264990: WebEngine crashes with segfault when not loaded through system classloader

matthiasblaesing opened the pull request:
The functions from FileSystemJava are called from different threads the
root problem manifests because the JNI FindClass function behaves
differently when called from a context that is the ancestor of a java
frame compared to when called in isolation.

A segmentation fault is observed when local storage of a webview is
accessed. At that time a new native thread is spun up and that sets up
the local storage, by calling into the JVM via
WTF::FileSystem::makeAllDirectories. At that point GetFileSystemClass is
invoked to get a referenc to the java implementation of the FileSystem.
As this is is called from a new native thread (no java context
available), JNI uses the system classloader to locate the class. This
fails if the JavaFX modules are not on the boot module/class path.

Instead on relying on fetching the class reference everytime it is
needed, this change fetches it once when the JavaFX library is loaded
and stores it in the WTF namespace.

In addition to this it was observed, that there is no attachment to the
JVM done when calling into the filesystem. No fault was observed, but
the JNI specs indicate, that the JNIEnv interface is only valid when
attached.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264990](https://bugs.openjdk.java.net/browse/JDK-8264990): WebEngine crashes with segfault when not loaded through system classloader


### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Arun Joseph](https://openjdk.java.net/census#ajoseph) (@arun-Joseph - Committer)
 * [Johan Vos](https://openjdk.java.net/census#jvos) (@johanvos - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/458/head:pull/458` \
`$ git checkout pull/458`

Update a local copy of the PR: \
`$ git checkout pull/458` \
`$ git pull https://git.openjdk.java.net/jfx pull/458/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 458`

View PR using the GUI difftool: \
`$ git pr show -t 458`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/458.diff">https://git.openjdk.java.net/jfx/pull/458.diff</a>

</details>

kevinrushforth commented:
/reviewers 2

matthiasblaesing commented:
The issue reference is the ID I got after I requested a bug to be created - I'm waiting to have it being opened up.

johanvos commented:
We have a concurrency issue here, as I created https://bugs.openjdk.java.net/browse/JDK-8264886 and a PR https://github.com/openjdk/jfx/pull/459 .
However, I recommend we proceed with this PR, as it also fixes the failure with the WebKitFileSystem class reference (which should indeed be done in the OnLoad). 
The fix looks good, I'll have a look at the tests.

kevinrushforth commented on tests/system/src/testapp7/java/mymod/myapp7/LocalStorageAccessWithModuleLayer.java:
```diff
@@ -0,0 +1,99 @@
+                        }
+                    }
+                });
+        webview.getEngine().load(LocalStorageAccessWithModuleLayer.class.getResource("/LocalStorageAccess.html").toExternalForm());
```
This fails to find the resource when I run the test via gradle. I recommend putting it in the same package as the class and removing the leading `/` (e.g., see the FXML resources in `testapp6` or `testscriptapp1`).

matthiasblaesing replied:
Can you tell me how you ran it? I ensured, that the test failed with a segfault before continuing with the fix. While I did that I indeed placed the file in the wrong directory, but currently I see no test failures locally. I run:

```
./gradlew  -PFULL_TEST=true -PUSE_ROBOT=true :systemTests:cleanTest :systemTests:test --tests test.com.sun.webkit.LocalStorageAccessTest 
```

to only invoke the new test. I'll move the file to the package, but I'd like to understand where I went wrong.

kevinrushforth replied:
That's how I ran it, too. I did it from a clean repo. This was on Windows, but shouldn't matter.

kevinrushforth replied:
I just tried it on Linux and get the same failure:

```
Caused by: java.lang.NullPointerException: Cannot invoke "java.net.URL.toExternalForm()" because the return value of "java.lang.Class.getResource(String)" is null
        at mymod/myapp7.LocalStorageAccessWithModuleLayer.start(LocalStorageAccessWithModuleLayer.java:92)
        at javafx.graphics/com.sun.javafx.application.LauncherImpl.lambda$launchApplication1$9(LauncherImpl.java:849)
        at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runAndWait$12(PlatformImpl.java:474)
        at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:447)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
        at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:446)
        at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
        at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
        at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$11(GtkApplication.java:277)
        ... 1 more
```

johanvos replied:
I get the same error. For some reason, the resource isn't found when it's on the root.
Moving it inside the app7 package work.

matthiasblaesing replied:
Thank you both for checking. I noticed, there were two copies of the HTML present in the `build` folder, which explains the strange behavior.

kevinrushforth commented on modules/javafx.web/src/main/native/Source/WTF/wtf/java/FileSystemJava.cpp:
```diff
@@ -246,15 +245,16 @@ PlatformFileHandle openFile(const String& path, FileOpenMode mode, FileAccessPer
     if (isHandleValid(handle)) {
-        JNIEnv* env = WTF::GetJavaEnv();
+        AttachThreadAsDaemonToJavaEnv autoAttach;
+    JNIEnv* env = autoAttach.env();
```
Minor: indentation

kevinrushforth commented on modules/javafx.web/src/main/native/Source/WTF/wtf/java/JavaEnv.cpp:
```diff
@@ -131,6 +132,17 @@ JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM* vm, void*)
+    // library is also used to load the target class, which is by construction
+    // the right one
+    static JGClass fileSystemRef(env->FindClass("com/sun/webkit/FileSystem"));
+    WTF::comSunWebkitFileSystem = fileSystemRef;
```
Can you `ASSERT` that the returned class is not null?

kevinrushforth commented on tests/system/src/testapp7/java/mymod/myapp7/LocalStorageAccessWithModuleLayer.java:
```diff
@@ -0,0 +1,99 @@
+        // 'Executed'. This indicates, that the second script block was reached
+        webview.getEngine().getLoadWorker().stateProperty().addListener(
+                new ChangeListener<State>() {
+                    public void changed(ObservableValue ov, State oldState, State newState) {
```
Suggestion: maybe use a lambda here to make the code more compact? (it's up to you whether to change it)

kevinrushforth commented on tests/system/src/testapp7/java/mymod/myapp7/LocalStorageAccessWithModuleLayerLauncher.java:
```diff
@@ -0,0 +1,83 @@
+
+        // Hack to get the classes of this programm into a module layer
+        List<Path> modulePaths = new ArrayList<>();
+        for(String workerPath: System.getProperty("module.path").split(File.pathSeparator)) {
```
Minor: missing space after the `for` and also before the `:`

kevinrushforth reviewed (commented):
The fix looks OK to me, although I'd like @arun-Joseph to comfirm.

The fix needs one change in how the resource (the `.html` file) is located, and I added a couple minor comments on the test. I'll test it after that.

kevinrushforth commented:
To follow up on my earlier comment, if I move `LocalStorageAccess.html` to `tests/system/src/testapp7/resources/mymod/myapp7/` and remove the `/` from the program that loads it, the test is able to load the resource. And I verified that the test fails without your fix and passes with your fix.

kevinrushforth commented:
The bug is now available in the JDK project in JBS:
https://bugs.openjdk.java.net/browse/JDK-8264990

matthiasblaesing commented:
@kevinrushforth @johanvos thank you both for review/checking. I pushed an update, that addresses the comments and updated the issue with the final issue id. I did not touch the comments on the original commits, so that history is kept intact (I understand, that skara bot will use the PR message as base for the final squashed commit and ignores the individual commits)

kevinrushforth commented:
>  I pushed an update, that addresses the comments and updated the issue with the final issue id.

You also need to adjust the title to match the bug title exactly.

> I did not touch the comments on the original commits, so that history is kept intact (I understand, that skara bot will use the PR message as base for the final squashed commit and ignores the individual commits)

Yes, this is correct. And we prefer contributors to do just what you did for incremental changes. Thanks.

matthiasblaesing commented:
I'm currently rerunning tests and preparing and update - further analysis showed, that thread attachment is handled in `WebCore::StorageThread::threadEntryPoint()`:

https://github.com/openjdk/jfx/blob/808b1078f762a923bd5e74298daffeb88ed108c2/modules/javafx.web/src/main/native/Source/WebKitLegacy/Storage/StorageThread.cpp#L76-L87

The analysis was triggered by a request from @neilcsmith-net on the netbeans mailing list. He pointed out, that the fault should have happened earlier if the the thread indeed was not attached. In that case  `GetEnv` is required to return `NULL`. From the [doc](https://docs.oracle.com/en/java/javase/11/docs/specs/jni/invocation.html#getenv):

> If the current thread is not attached to the VM, sets *env to NULL, and returns JNI_EDETACHED. If the specified version is not supported, sets *env to NULL, and returns JNI_EVERSION. Otherwise, sets *env to the appropriate interface, and returns JNI_OK.

kevinrushforth commented:
That's good since it will simplify the fix to the root cause, which is that looking up the class fails depending on the ClassLoader used to load the JavaFX modules.

johanvos commented:
I understand the crash would be at a different point, but it would still be a crash, I think? That is, if the method handlers are correctly registered via the OnLoad function, and later a new native thread tries to access a function that is not yet attached, that will still crash, I believe. (at least, the GetEnv will return NULL and that result is used later).
So while the 2 issues are not related, I think there are really 2 issues.

neilcsmith-net commented:
@johanvos a quick comment as Matthias linked me on this.  The reason I suggested re-looking is that there is no evidence of thread attachment being behind this problem in the discussion that happened on the NetBeans list, in fact this problem couldn't happen if the thread was unattached.  It depends if that discussion was the only basis for assuming an issue with thread attachment exists?

matthiasblaesing commented:
> I understand the crash would be at a different point, but it would still be a crash, I think? That is, if the method handlers are correctly registered via the OnLoad function, and later a new native thread tries to access a function that is not yet attached, that will still crash, I believe. (at least, the GetEnv will return NULL and that result is used later).
> So while the 2 issues are not related, I think there are really 2 issues.

The question is: Is there a second issue? If all threads used in webkit are attached to the VM, any thread can directly fetch an `JNIEnv` instance from the `vm` interface and invoke methods. In this case the thread is correctly attached.

The problem here is/was, that the code path is not regularly exercised. My feeling is, that it is common, that the JavaFX modules are loaeded from the module path and loading JavaFX from a module layer / classpath (yes I know that this is not supported) is mostly untested.

kevinrushforth commented:
> The question is: Is there a second issue?

If we don't have an example where a thread that isn't attached to the JVM already calls into these APIs, then it seems like the second issue is just hypothetical.

In any case, it seems best to separate it out.

@johanvos Do you have a case where it is actually failing as a result of a thread not being attached?

arun-Joseph commented on tests/system/src/test/java/test/com/sun/webkit/LocalStorageAccessTest.java:
```diff
@@ -0,0 +1,75 @@
+
+/**
+ * @test
+ * @bug 9069811
```
The changes looks good. The bug id is not modified at this point.

matthiasblaesing replied:
Fixed, sorry I missed it after the issue was made public.

johanvos commented:
> @johanvos Do you have a case where it is actually failing as a result of a thread not being attached?

No, I don't have a usecase for that. Hence, as long as there is no trace that can lead to this, I agree this (calling the Attach/Detach) doesn't need to be added.

kevinrushforth reviewed (approved):
Looks good. Tested on all three platforms. I verified (on both Mac and Linux) that the new test fails without the fix and passes with the fix.

arun-Joseph reviewed (approved):
The test and changes looks good.

matthiasblaesing commented:
/integrate

matthiasblaesing commented:
@arun-Joseph @kevinrushforth thank you both for review. Would one of you be kind enough to sponsor this change set? Thank you.

johanvos commented:
/sponsor

matthiasblaesing commented:
@johanvos thanks for review and sponsoring!
